cobra/util/solver.py

```diff
@@ -82,45 +82,47 @@ def set_objective(model, value, additive=False):
     additive : bool
         If true, add the terms to the current objective, otherwise start with
         an empty objective.
+
+    check_context : bool
+        Whether or not to skip the model's context manager. Internal use only.
     """
-    by_objective = isinstance(value,
-                              (sympy.Basic, model.solver.interface.Objective))
-
-    not_supported = (
-        additive and (not model.objective.is_Linear and not by_objective))
-    if not_supported:
-        raise ValueError('can only update non-linear objectives additively '
-                         'using object of class '
-                         'model.solver.interface.Objective, not %s' %
-                         type(value))
     reverse_value = None
-    if by_objective:
-        if not additive:
-            if isinstance(value, sympy.Basic):
-                value = model.solver.interface.Objective(value, sloppy=False)
-            model.solver.objective = value
-        else:
-            if isinstance(value, model.solver.interface.Objective):
-                value = value.expression
-            model.solver.objective += value
```

Comment:
Or substitute it for a reset/zero args that zeros the objective before applying the requested changes. This can simply be implemented with a 

```Python
model.solver.objective = model.solver.interface.Objective(S.Zero)
```

at the beginning of the function.